lib/services/parser.go

```diff
@@ -731,136 +732,69 @@ func newParserForIdentifierSubcondition(ctx RuleContext, identifier string) (pre
-				default:
-					key := fields[1]
-					val, ok := resource.GetLabel(key)
-					if ok {
-						return label{key: key, value: val}, nil
-					}
-					return label{}, nil
-				}
-
-			case ResourceNameIdentifier:
-				if len(fields) > 1 {
-					return nil, trace.BadParameter("only one field are supported with identifier %q, got %d: %v", ResourceNameIdentifier, len(fields), fields)
-				}
-
+func NewResourceParser(expression string) (typical.Expression[types.ResourceWithLabels, bool], error) {
+	parser, err := typical.NewCachedParser[types.ResourceWithLabels, bool](typical.ParserSpec[types.ResourceWithLabels]{
+		Variables: map[string]typical.Variable{
+			"resource.metadata.labels": typical.DynamicVariable(func(r types.ResourceWithLabels) (map[string]string, error) {
+				return r.GetStaticLabels(), nil
+			}),
+			"resource.metadata.name": typical.DynamicVariable(func(r types.ResourceWithLabels) (string, error) {
+				return r.GetName(), nil
+			}),
+			"labels": typical.DynamicMapFunction(func(r types.ResourceWithLabels, key string) (string, error) {
+				val, _ := r.GetLabel(key)
+				return val, nil
+			}),
+			"name": typical.DynamicVariable(func(r types.ResourceWithLabels) (string, error) {
 				// For nodes, the resource "name" that user expects is the
-				// nodes hostname, not its UUID. Currently for other resources,
+				// nodes hostname, not its UUID. Currently, for other resources,
 				// the metadata.name returns the name as expected.
-				if server, ok := resource.(types.Server); ok {
+				if server, ok := r.(types.Server); ok {
 					return server.GetHostname(), nil
 				}
 
-				return resource.GetName(), nil
-			case ResourceIdentifier:
-				return predicate.GetFieldByTag(resource, teleport.JSON, fields[1:])
-			default:
-				return nil, trace.NotFound("identifier %q is not defined", strings.Join(fields, "."))
-			}
+				return r.GetName(), nil
+			}),
 		},
-		GetProperty: func(mapVal, keyVal interface{}) (interface{}, error) {
-			r, ok := mapVal.(types.ResourceWithLabels)
-			if !ok {
-				return GetStringMapValue(mapVal, keyVal)
-			}
-
-			key, ok := keyVal.(string)
-			if !ok {
-				return nil, trace.BadParameter("only string keys are supported")
+		Functions: map[string]typical.Function{
+			"hasPrefix": typical.BinaryFunction[types.ResourceWithLabels, string, string, bool](func(s, suffix string) (bool, error) {
+				return strings.HasPrefix(s, suffix), nil
+			}),
+			"equals": typical.BinaryFunction[types.ResourceWithLabels, string, string, bool](func(a, b string) (bool, error) {
+				return strings.Compare(a, b) == 0, nil
+			}),
+			"search": typical.UnaryVariadicFunctionWithEnv(func(r types.ResourceWithLabels, v ...string) (bool, error) {
+				return r.MatchSearch(v), nil
+			}),
+			"exists": typical.UnaryFunction[types.ResourceWithLabels, string, bool](func(value string) (bool, error) {
+				return value != "", nil
+			}),
```

Comment:
nit: I guess it's a matter of preference, but all these types except for the env can be automatically inferred, and I tend to find it more confusing to see a list of random types that has to be maintained
```suggestion
			"hasPrefix": typical.BinaryFunction[types.ResourceWithLabels](func(s, suffix string) (bool, error) {
				return strings.HasPrefix(s, suffix), nil
			}),
			"equals": typical.BinaryFunction[types.ResourceWithLabels](func(a, b string) (bool, error) {
				return strings.Compare(a, b) == 0, nil
			}),
			"search": typical.UnaryVariadicFunctionWithEnv(func(r types.ResourceWithLabels, v ...string) (bool, error) {
				return r.MatchSearch(v), nil
			}),
			"exists": typical.UnaryFunction[types.ResourceWithLabels](func(value string) (bool, error) {
				return value != "", nil
			}),
```